### PR TITLE
Implements PGOBBAddrMap in Object and ObjectYAML with tests [1/5]

### PR DESCRIPTION
**Pull request by red1bluelost**
A part of [RFC - PGO Accuracy Metrics: Emitting and Evaluating Branch and Block Analysis](https://discourse.llvm.org/t/rfc-pgo-accuracy-metrics-emitting-and-evaluating-branch-and-block-analysis/73902).

This PR adds the PGOBBAddrMap data structure and implements encoding and decoding through Object and ObjectYAML along with associated tests.

## Design Alternative
The current implementation makes an effort to avoid modifying the original BBAddrMap data structure. This keeps the API the same as before for all pre-existing code and avoids unnecessary fields for regular BBAddrMap usage. It comes with the cost of some complexity in sharing encoding/decoding code.

Alternatively, we could just place all new fields into BBAddrMap. This makes the encoding/decoding code simpler. It has the cost of changing some behavior in the original API. It also adds unnecessary fields for anyone wanting to just use the regular BBAddrMap.

Please let us know if you feel that the alternate design of placing new fields into BBAddrMap is preferable.

## PR Series

The current code for PGOBBAddrMap to be upstreamed is split into five PRs:
1. Object and ObjectYAML - (this one) https://github.com/red1bluelost/llvm-project/tree/pgo-bb-addr-map--object
2. AsmPrinter - https://github.com/red1bluelost/llvm-project/pull/1 or https://github.com/red1bluelost/llvm-project/tree/pgo-bb-addr-map--asm-printer
3. llvm-readobj - https://github.com/red1bluelost/llvm-project/pull/2 or https://github.com/red1bluelost/llvm-project/tree/pgo-bb-addr-map--llvm-readobj
4. llvm-objdump - https://github.com/red1bluelost/llvm-project/tree/pgo-bb-addr-map--llvm-objdump
5. llvm obj2yaml - https://github.com/red1bluelost/llvm-project/tree/pgo-bb-addr-map--llvm-obj2yaml

If you would like to try testing PGOBBAddrMap locally on a program, PR-3 llvm-readobj is likely the minimum code needed to meaningfully use this feature.

### CODE REVIEW

**Comment by jh7370** [comment]
I've added @rlavaee as a reviewer, as they've done a lot of the work for the BB support in the various tools. I think it makes most sense for him to do the high-level review of this.

**Comment by rlavaee** [comment]
Haven't read the code yet, but this is my high-level comment:

> Alternatively, we could just place all new fields into BBAddrMap. This makes the encoding/decoding code simpler. It has the cost of changing some behavior in the original API. It also adds unnecessary fields for anyone wanting to just use the regular BBAddrMap.

We can distinguish between the ELF representation (SHT_LLVM_BB_ADDR_MAP) and the in-memory representation (`BBAddrMap` which is returned by `readBBAddrMap`). It's perfectly fine to add new fields to the ELF representation. This is why the `feature` and `version` attributes exist. This means we don't need to define a new section type and duplicate the shared encoding/decoding logic. However, it's a sensible idea to define a new `PGOBBAddrMap` class so we don't clobber the existing `BBAddrMap` with unnecessary fields. WDYT?

**Comment by red1bluelost** [comment]
> Haven't read the code yet, but this is my high-level comment:
> 
> > Alternatively, we could just place all new fields into BBAddrMap. This makes the encoding/decoding code simpler. It has the cost of changing some behavior in the original API. It also adds unnecessary fields for anyone wanting to just use the regular BBAddrMap.
> 
> We can distinguish between the ELF representation (SHT_LLVM_BB_ADDR_MAP) and the in-memory representation (`BBAddrMap` which is returned by `readBBAddrMap`). It's perfectly fine to add new fields to the ELF representation. This is why the `feature` and `version` attributes exist. This means we don't need to define a new section type and duplicate the encoding/decoding logic. However, it's a sensible idea to define a new `PGOBBAddrMap` class so we don't clobber the existing `BBAddrMap` with unnecessary fields. WDYT?

I see what you mean by sharing ELF section (SHT_LLVM_BB_ADDR_MAP) rather than having it separate (SHT_PGO_LLVM_BB_ADDR_MAP). We would use feature byte to indicate enabled analyses. I would not be opposed to doing this.

If we keep the two classes `PGOBBAddrMap` and `BBAddrMap`, I don't think that this would simplify the encoding/decoding logic that much. `PGOBBAddrMap` would be the same while `BBAddrMap` would either need to error if Feature is not zero or decode but ignore the extra data. I personally would lean towards having `BBAddrMap` ignore extra data since `PGO...` is a superset and I suspect ignoring could simplify 'trait design' complexity.

I'll give a try to combine and have BBAddrMap ignore extra data in the mean time.

**Comment by rlavaee** [comment]
> If we keep the two classes `PGOBBAddrMap` and `BBAddrMap`, I don't think that this would simplify the encoding/decoding logic that much. `PGOBBAddrMap` would be the same while `BBAddrMap` would either need to error if Feature is not zero or decode but ignore the extra data. I personally would lean towards having `BBAddrMap` ignore extra data since `PGO...` is a superset and I suspect ignoring could simplify 'trait design' complexity.

One way to do this is to have the decoding function return two structures `std::vector<BBAddrMap>` and `std::vector<PGOFrequencyMap>`. `PGOFrequencyMap` does not store any address information so we don't need to duplicate the base fields. The only drawback is that we need to define the mapping between the two data structures.

For decoding, we can inspect the feature field and populate `PGOFrequencyMap` if we need to. WDYT?

**Comment by red1bluelost** [comment]
> > If we keep the two classes `PGOBBAddrMap` and `BBAddrMap`, I don't think that this would simplify the encoding/decoding logic that much. `PGOBBAddrMap` would be the same while `BBAddrMap` would either need to error if Feature is not zero or decode but ignore the extra data. I personally would lean towards having `BBAddrMap` ignore extra data since `PGO...` is a superset and I suspect ignoring could simplify 'trait design' complexity.
> 
> One way to do this is to have the decoding function return two structures `std::vector<BBAddrMap>` and `std::vector<PGOFrequencyMap>`. `PGOFrequencyMap` does not store any address information so we don't need to duplicate the base fields. The only drawback is that we need to define the mapping between the two data structures.
> 
> For decoding, we can inspect the feature field and populate `PGOFrequencyMap` if we need to. WDYT?

I see. I think that should be better. Using the "PGOFrequencyMap" will definitely help simplify Object's decode/encode. ObjectYAML is the harder area to tackle. I have one more idea to try so that we don't need new fields in yaml::BBAddrMapEntry, but if it fails, it might be cleanest to duplicate the fields just in yaml.

I think mapping should be fine. I see two designs.
1. We could just keep the sizes identical to have one-to-one mapping. The PGOFrequencyMap would contain FuncEntryCount and a vector of BBFreq and BrProbs. In the case where a function is cold or empty, we could just leave the block vector empty.
2. We could include the function address in PGOFrequencyMap and not make cold or empty function. I'm not sure if there on concerns with unique addresses.

I'll try building the first designs.

**Comment by red1bluelost** [comment]
I went ahead and merged into this PR the design changes suggested @rlavaee. (I called it PGOAnalysisMap instead of PGOFrequencyMap). Using the same ELF section with the feature guard does help to simplify and shorten the code. If the user does not request the PGOAnalysisMap when decoding the BBAddrMap section then the code just ignores any extra data.

File: llvm/lib/ObjectYAML/ELFYAML.cpp
Comment (jh7370) [inline llvm/lib/ObjectYAML/ELFYAML.cpp:1845]: Are these fields both REALLY required, or could reasonable defaults (including empty strings, 0 etc) be used?

**Review by jh7370 — COMMENTED** [review]
Some comments - I haven't done a particularly detailed review.

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp]: Must've been my LSP automatically adding them. Removed since unnecessary.

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp:663]: The `ULEBSizeErr` is not the same as the Cursor's error. It is the error if the value is between the requested max number of bits to 64-bit while the cursor takes all the other ULEB128 errors.

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp:658]: That could work but it looks a bit ugly since formats weird and you need bool_constant. Since this method is just a static helper, the static_assert is cleaner and avoids accidental misuse.

File: llvm/lib/ObjectYAML/ELFYAML.cpp
Comment (red1bluelost) [reply llvm/lib/ObjectYAML/ELFYAML.cpp:1845]: If you are listing the successors in the YAML, these two fields are essential. It's optional to have the list of successor but not optional for those fields.
Maybe in the future if SuccessorEntry gets more fields then optional would become worth it.

**Review by rlavaee — COMMENTED** [review]
Thanks for addressing my concerns.

I wonder if there is any benefit in emitting the PGOAnalysis BB entries next to their BBAddrMap counterparts in the section. I see we are using the Metadata field to emit the successor type (used primarily for succinct decoding/encoding), but we probably shouldn't introduce this kind of coupling.

Here is an alternative:
Emit the entire PGOAnalysisMap entry of each function right after its BBAddrMap (NBlocks and Feature are shared). Define and use your own fancy encoding for successor type and count.

File: llvm/lib/Object/ELF.cpp
Comment (jh7370) [reply llvm/lib/Object/ELF.cpp:658]: I'm not sure if you're doing the same thing I imagined. I played around with your code, and the template signature just becomes:
```
template <typename IntTy,
          typename = std::enable_if_t<std::is_unsigned_v<IntTy> &&
                                      (std::numeric_limits<IntTy>::radix == 2)>>
```
Which doesn't seem weirdly formatted to me and doesn't use `bool_constant`? I think `enable_if` (or any other SFINAE approach) is the more canonical way in C++ of preventing use of a template function rather than a `static_assert`.

But anyway, is the `radix` part actually needed? What case would it suppress that you think is potentially going to happen? What would be the issue if the function were to be used with such an integer type (if the static_assert/enable_if weren't present)?

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp:665]: Yes, I this is already tested in the regular BBAddrMap tests. Found at ELFObjectFileTest.cpp:564

**Comment by red1bluelost** [comment]
> Thanks for addressing my concerns.
> 
> I wonder if there is any benefit in emitting the PGOAnalysis BB entries next to their BBAddrMap counterparts in the section. I see we are using the Metadata field to emit the successor type (used primarily for succinct decoding/encoding), but we probably shouldn't introduce this kind of coupling.
> 
> Here is an alternative: Emit the entire PGOAnalysisMap entry of each function right after its BBAddrMap (NBlocks and Feature are shared). Define and use your own fancy encoding for successor type and count.

I'll try that out. Without packing into Metadata there is no space saved for Branch Probability. But that probably doesn't matter much.

**Comment by red1bluelost** [comment]
> Thanks for addressing my concerns.
> 
> I wonder if there is any benefit in emitting the PGOAnalysis BB entries next to their BBAddrMap counterparts in the section. I see we are using the Metadata field to emit the successor type (used primarily for succinct decoding/encoding), but we probably shouldn't introduce this kind of coupling.
> 
> Here is an alternative: Emit the entire PGOAnalysisMap entry of each function right after its BBAddrMap (NBlocks and Feature are shared). Define and use your own fancy encoding for successor type and count.

Done. It now encoded PGOAnalysisMap after each BBAddrMap function if it was enabled in Features.

File: llvm/lib/Object/ELF.cpp
Comment (jh7370) [inline llvm/lib/Object/ELF.cpp:13]: Did this sneak back in in your latest round of changes?

**Review by jh7370 — COMMENTED** [review]
Thanks for the updates. No more comments from me, barring one possible nit.

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp]: Yep

File: llvm/unittests/Object/ELFObjectFileTest.cpp
Comment (rlavaee) [inline llvm/unittests/Object/ELFObjectFileTest.cpp:783]: "versions that are too old"

File: llvm/unittests/Object/ELFObjectFileTest.cpp
Comment (rlavaee) [inline llvm/unittests/Object/ELFObjectFileTest.cpp:838]: Not sure if thiis is right. Feature value is the same as the one above.

File: llvm/lib/Object/ELF.cpp
Comment (red1bluelost) [reply llvm/lib/Object/ELF.cpp]: I like that idea, implemented.

File: llvm/lib/Object/ELF.cpp
Comment (rlavaee) [inline llvm/lib/Object/ELF.cpp:720]: Better to just check for error in Cur and break instead, similar to L713.

**Review by rlavaee — APPROVED** [review]
Thank you for the updates. Only one comment from me.

**Comment by rlavaee** [comment]
Also, please add the "[SHT_LLVM_BB_ADDR_MAP]"  tag to the commit message.

File: llvm/unittests/Object/ELFObjectFileTest.cpp
Comment (jh7370) [inline llvm/unittests/Object/ELFObjectFileTest.cpp:784]:
```suggestion
  // Check that we can detect unsupported versions that are too old.
```
Nit.

File: llvm/unittests/Object/ELFObjectFileTest.cpp
Comment (jh7370) [inline llvm/unittests/Object/ELFObjectFileTest.cpp:869]: Nit.
```suggestion
// Test for the ELFObjectFile::readBBAddrMap API with PGOAnalysisMap.
```

**Review by jh7370 — APPROVED** [review]
Two final nitpicks from me (I didn't check to see if there were other cases of the same issue, but if there are, please fix them.

**Comment by rlavaee** [comment]
Would you please wait until I merge https://github.com/llvm/llvm-project/pull/74107? This should allow removing some boilerplate code from your PR as well.

**Pull request merged**
